numba/parfors/parfor.py

```diff
@@ -4064,28 +4072,111 @@ def remove_duplicate_definitions(blocks, nameset):
+                    if (all([x.name in indices or
+                        x.name not in derived_from_indices for x in ind_seq])):
+                        new_index_positions = [x.name in indices for x in ind_seq]
+                        # Make sure that we aren't accessing a given array with
+                        # different indices in a different order.
+                        if add_check_position(new_index_positions, array_accessed):
+                            return True
+                    else:
+                        return True
+                else:
+                    # Don't know how the index tuple is built so
+                    # have to assume fusion can't happen.
+                    return True
+            else:
+                # Should be for 1D array getitems.
+                if stmt_index.name in indices:
+                    if add_check_position((True,), array_accessed):
+                        return True
+                elif stmt_index.name in derived_from_indices:
+                    # If we ever index an array with something calculated
+                    # from an index then no fusion.
+                    return True
+                else:
+                    if add_check_position((False,), array_accessed):
+                        return True
+        else:
+            return True
+        return False
+
     for b in parfor.loop_body.values():
         for stmt in b.body:
             # GetItem/SetItem nodes are fine since can't have expression inside
             # and only simple indices are possible
             if isinstance(stmt, (ir.SetItem, ir.StaticSetItem)):
+                if check_index(stmt.index, stmt.target.name):
+                    return True, index_positions, indexed_arrays, non_indexed_arrays
                 continue
             # tuples are immutable so no expression on parfor possible
-            if isinstance(stmt, ir.Assign) and isinstance(stmt.value, ir.Expr):
-                op = stmt.value.op
-                if op in ['build_tuple', 'getitem', 'static_getitem']:
-                    continue
-            # other statements can have potential violations
-            if not indices.isdisjoint(stmt.list_vars()):
-                dprint("has_cross_iter_dep found", indices, stmt)
-                return True
-    return False
+            if isinstance(stmt, ir.Assign):
+                if isinstance(stmt.value, ir.Var):
+                    if stmt.value.name in indices:
+                        indices.add(stmt.target.name)
+                        continue
+                elif isinstance(stmt.value, ir.Expr):
+                    op = stmt.value.op
+                    if op in ['getitem', 'static_getitem']:
+                        if not check_index(stmt.value.index, stmt.value.value.name):
+                            continue
+
+                        return True, index_positions, indexed_arrays, non_indexed_arrays
```

Comment:
This expression evaluates the same as the `check_index` under `GetItem/SetItem` but is written as `if not`, perhaps make both the same for consistency?

Reply:
Done.